### PR TITLE
1940/2091 - Data Table missing types and incorrect slot rendering

### PR DESCRIPTION
**Pull request by GCHQ-Developer-741**
## Summary of the changes
- Added export to canary-web-components interface to include missing types relating to data tables. Couldn't add to components.d.ts as they kept getting purged by stencil.
- Added missing return types in ic-data-table that was causing linting errors
- Added helper function to add placeholder values that represent slotted content to data in the correct column order rather than at the end of the object, so that it can render slotted content regardless of which order the columns are in
- Updated test screenshots/data to show the change working

## Related issue
#2091 

### General 

- [x] Changes to docs package checked and committed.
- [x] All acceptance criteria reviewed and met. 

### Testing

- [x] Relevant unit tests and visual regression tests added. 
- [x] Visual testing against Figma component specification completed. 
- [x] Playground stories in React Storybook up to date, with any prop changes and additions addressed. 
- [x] Compare performance of modified components against develop using Performance addon in React Storybook.

### Accessibility 

- [x] Accessibility Insights FastPass performed.
- [x] A11y unit test added and yields no issues.
- [x] A11y plug-in on Storybook yields no issues. 
- [x] Manual screen reader testing performed using NVDA and VoiceOver. 
- [x] Manual keyboard testing for keyboard controls and logical focus order. 
- [x] Correct roles used and ARIA attributes used correctly where required. 
- [x] Logical heading structure is maintained, and the HTML elements used for headings can be changed to fit within the wider page structure. 

### Resize/zoom behaviour 

- [x] Page can be zoomed to 400% with no loss of content. 
- [x] Screen magnifier used with no issues. 
- [x] Text resized to 200% with no loss of content.
- [x] Text spacing increased as per the [WCAG 1.4.12 success criterion](https://www.w3.org/TR/WCAG21/#text-spacing) with no loss of content.

### System modes

- [x] Browser setting 'prefers reduced motion' tested. No animations or motion visible whilst this setting is on. 
- [x] Windows High Contrast mode tested with no loss of content.
- [x] System light and dark mode tested with no loss of content.
- [x] Browser support tested (Chrome, Safari, Firefox and Edge). 

### Testing content extremes

- [x] Min/max content examples tested with no loss of content or overflow. 
- [x] All prop combinations work without issue. 
- [x] Tested for FOUC (Flash of Unstyled Content) in both SSR (Server-Side Rendering) and SSG (Static Site Generation) settings.
- [x] Controlled and uncontrolled input components tested.
- [x] Props/slots can be updated after initial render.

### CODE REVIEW

**Review by jd239 — CHANGES_REQUESTED** [review]
Cheers for looking into this. A few small comments.

**Comment by jd239** [comment]
Just tried creating a table with two columns of custom elements and it didn't appear correctly. The example below is using the React slots

```jsx
    <IcDataTable
      caption="Links and Elements in data"
      columns={COLS_ELEMENTS}
      data={DATA_REACT_ELEMENTS}
    >
      {DATA_REACT_ELEMENTS.map((_, index) => (
        <IcButton
          key={index}
          variant="destructive"
          slot={`actions-${index}`}
          onClick={() => console.log("Delete")}
        >
          Delete
        </IcButton>
      ))}
      {DATA_REACT_ELEMENTS.map((_, index) => (
        <IcButton
          key={index}
          variant="tertiary"
          slot={`custom-${index}`}
          onClick={() => console.log("Delete")}
        >
          Delete
        </IcButton>
      ))}
    </IcDataTable>
```

```js
export const COLS_ELEMENTS: IcDataTableColumnObject[] = [
  {
    key: "actions",
    title: "Actions",
    dataType: "element",
    columnAlignment: { horizontal: "center" },
  },
  {
    key: "firstName",
    title: "First name",
    dataType: "string",
  },
  {
    key: "lastName",
    title: "Last name",
    dataType: "string",
  },
  {
    key: "custom",
    title: "Custom",
    dataType: "element",
  },
  {
    key: "age",
    title: "Age",
    dataType: "number",
  },
  {
    key: "jobTitle",
    title: "Job title",
    dataType: "string",
  },
  {
    key: "address",
    title: "Address",
    dataType: "address",
  },
];

```

File: packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx [hunk 1]
```diff
@@ -1258,17 +1258,23 @@ export class DataTable {
...
+     * `addDataToPosition` used to add the element in the correct column order.
      */
-    this.columns.forEach(({ key }) => {
-      data.forEach((row, rowIndex) => {
-        const cellSlotName = `${key}-${rowIndex}`;
-        if (isSlotUsed(this.el, cellSlotName)) {
-          row[key] = getSlotContent(this.el, cellSlotName);
-        }
-      });
+    const organisedSlotData = data.map((row, rowIndex) => {
```
[the remark this answers is not in the document]
Comment (GCHQ-Developer-741) [reply packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx:1263]: Will update var name

**Comment by GCHQ-Developer-741** [comment]
> Just tried creating a table with two columns of custom elements and it didn't appear correctly. The example below is using the React slots
> 
> ```js
>     <IcDataTable
>       caption="Links and Elements in data"
>       columns={COLS_ELEMENTS}
>       data={DATA_REACT_ELEMENTS}
>     >
>       {DATA_REACT_ELEMENTS.map((_, index) => (
>         <IcButton
>           key={index}
>           variant="destructive"
>           slot={`actions-${index}`}
>           onClick={() => console.log("Delete")}
>         >
>           Delete
>         </IcButton>
>       ))}
>       {DATA_REACT_ELEMENTS.map((_, index) => (
>         <IcButton
>           key={index}
>           variant="tertiary"
>           slot={`custom-${index}`}
>           onClick={() => console.log("Delete")}
>         >
>           Delete
>         </IcButton>
>       ))}
>     </IcDataTable>
> ```
> 
> ```js
> export const COLS_ELEMENTS: IcDataTableColumnObject[] = [
>   {
>     key: "actions",
>     title: "Actions",
>     dataType: "element",
>     columnAlignment: { horizontal: "center" },
>   },
>   {
>     key: "firstName",
>     title: "First name",
>     dataType: "string",
>   },
>   {
>     key: "lastName",
>     title: "Last name",
>     dataType: "string",
>   },
>   {
>     key: "custom",
>     title: "Custom",
>     dataType: "element",
>   },
>   {
>     key: "age",
>     title: "Age",
>     dataType: "number",
>   },
>   {
>     key: "jobTitle",
>     title: "Job title",
>     dataType: "string",
>   },
>   {
>     key: "address",
>     title: "Address",
>     dataType: "address",
>   },
> ];
> ```

I think is due to the fact that when it looks through the columns to see which contain slotted content, it stops once it finds the first one. Will adjust this. Think the whole thing does need an adjustment to add multiple new values to the object rather than one. Cheers for finding this

**Review by jd239 — APPROVED** [review]

**Review by GCHQ-Developer-299 — APPROVED** [review]

**Pull request merged**